File: api/src/domain/entities/Playlist.js
'use strict';

const BaseEntity = require('./BaseEntity');

/**
 * Playlist Entity
 * Playlist domain modeli
 */
class Playlist extends BaseEntity {
    constructor(data = {}) {
        super(data);
        this.path = data.path || '';
        this.name = data.name || '';
        this.type = data.type || ''; // 'm3u' veya 'vdjfolder'
        this.track_count = data.track_count || 0;
        this.updated_at = data.updated_at || new Date().toISOString();
    }

    /**
     * M3U playlist mi?
     */
    isM3U() {
        return this.type === 'm3u';
    }

    /**
     * VDJ Folder playlist mi?
     */
    isVDJFolder() {
        return this.type === 'vdjfolder';
    }

    /**
     * Playlist boş mu?
     */
    isEmpty() {
        return this.track_count === 0;
    }

    /**
     * Track sayısını güncelle
     */
    updateTrackCount(count) {
        this.track_count = count;
        this.updated_at = new Date().toISOString();
    }

    /**
     * Validation
     */
    isValid() {
        return this.path && this.name && ['m3u', 'vdjfolder'].includes(this.type);
    }

    /**
     * Static factory method
     */
    static fromFile(filePath, playlistType) {
        const path = require('path');
        return new Playlist({
            path: filePath,
            name: path.basename(filePath, path.extname(filePath)),
            type: playlistType,
            track_count: 0
        });
    }
}

module.exports = Playlist;

File: xcode/PlaylistOrganizer/PlaylistOrganizer/Models/Playlist.swift
//
//  Playlist.swift
//  PlaylistOrganizer
//
//  Created by Koray Önür on 22.10.2025.
//

import Foundation

struct Playlist: Identifiable, Equatable, Codable {
    let id: Int
    let path: String
    let name: String
    let type: PlaylistType
    let trackCount: Int
    let createdAt: Date
    let updatedAt: Date
    
    // MARK: - UI Properties (temporary)
    var isSelected: Bool = false
    var isExpanded: Bool = false
    var children: [Playlist] = []
    var isChild: Bool = false
    
    // MARK: - Computed Properties
    
    /// M3U playlist mi?
    var isM3U: Bool {
        return type == .m3u
    }
    
    /// VDJ Folder playlist mi?
    var isVDJFolder: Bool {
        return type == .vdjfolder
    }
    
    /// Playlist boş mu?
    var isEmpty: Bool {
        return trackCount == 0
    }
    
    // MARK: - Validation
    
    /// Playlist geçerli mi?
    var isValid: Bool {
        return !path.isEmpty && !name.isEmpty && type != .unknown
    }
    
    // MARK: - Equatable
    
    static func == (lhs: Playlist, rhs: Playlist) -> Bool {
        return lhs.id == rhs.id
    }
    
    // MARK: - Factory Methods
    
    /// Dosyadan Playlist oluştur
    static func fromFile(_ filePath: String, playlistType: PlaylistType) -> Playlist {
        let fileName = URL(fileURLWithPath: filePath).lastPathComponent
        let name = URL(fileURLWithPath: filePath).deletingPathExtension().lastPathComponent
        
        return Playlist(
            id: 0, // Will be set by database
            path: filePath,
            name: name,
            type: playlistType,
            trackCount: 0,
            createdAt: Date(),
            updatedAt: Date()
        )
    }
}

enum PlaylistType: String, Codable, CaseIterable {
    case m3u = "m3u"
    case vdjfolder = "vdjfolder"
    case unknown = "unknown"
}
